File: app/src/main/java/com/rifafauzi/moviecatalogue/viewmodel/ViewModelFactory.java
package com.rifafauzi.moviecatalogue.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.rifafauzi.moviecatalogue.di.Injection;
import com.rifafauzi.moviecatalogue.helper.repository.Repository;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private static volatile ViewModelFactory INSTANCE;

    private final Repository repository;

    private ViewModelFactory(Repository repository) {
        this.repository = repository;
    }

    public static ViewModelFactory getInstance() {
        if (INSTANCE == null) {
            synchronized (ViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ViewModelFactory(Injection.repository());
                }
            }
        }
        return INSTANCE;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass.isAssignableFrom(MovieViewModel.class)) {
            //noinspection unchecked
            return (T) new MovieViewModel(repository);
        } else if (modelClass.isAssignableFrom(TvShowViewModel.class)) {
            //noinspection unchecked
            return (T) new TvShowViewModel(repository);
        }

        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }

}

File: app/src/main/java/com/rifafauzi/moviecatalogue/viewmodel/ViewModelFactory.kt
package com.rifafauzi.moviecatalogue.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import com.rifafauzi.moviecatalogue.di.Injection
import com.rifafauzi.moviecatalogue.helper.repository.Repository

@Suppress("UNCHECKED_CAST")
class ViewModelFactory private constructor(private val repository: Repository) : ViewModelProvider.NewInstanceFactory() {

    override fun <T : ViewModel> create(factoryClass: Class<T>) :T{
        return  when{
            factoryClass.isAssignableFrom(MovieViewModel::class.java) -> MovieViewModel(repository) as T
            factoryClass.isAssignableFrom(TvShowViewModel::class.java) -> TvShowViewModel(repository) as T
            else -> throw IllegalArgumentException("Unknown ViewModel: " + factoryClass.name)
        }
    }

    companion object {
        @Volatile
        private var INSTANCE: ViewModelFactory? = null

        fun getInstance(): ViewModelFactory? {
            if (INSTANCE == null) {
                synchronized(ViewModelFactory::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = Injection.repository()?.let { ViewModelFactory(it) }
                    }
                }
            }
            return INSTANCE
        }

    }

}
